bin/main.go

```diff
@@ -38,118 +37,106 @@ import (
-		},
-	})
-	if err != nil {
-		log.Warnf("Consul: %s", err)
-	}
-}
-
 func init() {
 	flag.Parse()
 	if *v {
 		config.PrintSinkerInfo()
 		os.Exit(0)
 	}
-	ip, port = parseAddr(*httpAddr)
+	ip = util.GetOutboundIP().String()
+	*httpPort = util.GetSpareTCPPort(ip, *httpPort)
+	selfAddr = fmt.Sprintf("%s:%d", ip, *httpPort)
 }
 
-// GenTasks generate the tasks via config
-func GenTasks(cfg *config.Config) (res []*task.Service) {
-	res = make([]*task.Service, 0, len(cfg.Tasks))
-	for _, taskCfg := range cfg.Tasks {
-		ck := output.NewClickHouse(taskCfg)
-		pp := parser.NewParserPool(taskCfg.Parser, taskCfg.CsvFormat, taskCfg.Delimiter, []string{taskCfg.LayoutDate, taskCfg.LayoutDateTime, taskCfg.LayoutDateTime64})
-		var inputer input.Inputer
-		if taskCfg.Kafka != "" {
-			inputer = input.NewInputer(taskCfg.KafkaClient)
-		}
-		taskImpl := task.NewTaskService(inputer, ck, taskCfg, pp)
-		res = append(res, taskImpl)
+// GenTask generate a task via config
+func GenTask(cfg *config.Config, taskName string) (taskImpl *task.Service) {
+	taskCfg := cfg.Tasks[taskName]
+	ck := output.NewClickHouse(cfg, taskName)
+	pp := parser.NewParserPool(taskCfg.Parser, taskCfg.CsvFormat, taskCfg.Delimiter, []string{taskCfg.LayoutDate, taskCfg.LayoutDateTime, taskCfg.LayoutDateTime64})
+	var inputer input.Inputer
+	if taskCfg.Kafka != "" {
+		inputer = input.NewInputer(taskCfg.KafkaClient)
 	}
+	taskImpl = task.NewTaskService(inputer, ck, pp, cfg, taskName)
 	return
 }
 
 func main() {
+	log.SetFormatter(&log.TextFormatter{
+		FullTimestamp: true,
+	})
 	if err := agent.Listen(agent.Options{}); err != nil {
-		log.Critical(err)
-	}
-	consulConfig := api.DefaultConfig()
-	consulConfig.Address = *consulAddr
-	consulClient, _ := api.NewClient(consulConfig)
-	consulAgent := consulClient.Agent()
-
-	if *consulRegister {
-		serviceRegister(consulAgent)
-		defer func() {
-			log.Debug("Consul: de-register service")
-			err := consulAgent.ServiceDeregister(appIDStr)
-			if err != nil {
-				log.Warnf("Consul: %s", err)
-			}
-		}()
+		log.Fatalf("%+v", err)
 	}
 
 	app.Run("clickhouse_sinker", func() error {
-		config.SetConfigDir(*cfgDir)
-		cfg := config.GetConfig()
-		runner = NewSinker(cfg)
+		var rcm config.RemoteConfManager
+		var properties map[string]interface{}
+		if *consulRegister {
+			rcm = &config.ConsulConfManager{}
+			properties = make(map[string]interface{})
+			properties["consulAddr"] = *consulAddr
+			properties["deregisterCriticalServiceAfter"] = *consulDeregisterCriticalServiceAfter
+			properties["ip"] = ip
+			properties["port"] = *httpPort
+		} else if *nacosRegister {
+			rcm = &config.NacosConfManager{}
+			properties = make(map[string]interface{})
+			properties["serverAddrs"] = *nacosAddr
+			properties["namespaceId"] = *nacosNamespaceID
+			properties["username"] = *nacosUsername
+			properties["password"] = *nacosPassword
+			properties["group"] = *nacosGroup
+			properties["ip"] = ip
+			properties["port"] = *httpPort
+		}
+		if rcm != nil {
```

Comment:
Local config can also be `ConfManager `, because we can introduce `fsnotify`

Reply:
I agree it's doable.